### DISCUSSION TITLE
Synthetic monitoring solution

### DISCUSSION THREAD

**Discussion by tristaZero** [SphereEx/CyborgFlow#4] [Ideas]
Hi community,

Good news here!

Apache APISIX, Apache Skywalking, and Apache ShardingSphere are collaborating on a synthetic monitoring solution. 🎉 
This solution is to help users auto-mark test data, auto-route the data to production and shadow databases, and monitor the online traffic through all the APPs and databases **ONLINE**.  At last, the integrated tracing, metrics will be drawn on UI.

Now we are sailing out!

#### Apache APISIX
👉🏻 https://github.com/apache/apisix

#### Apache Skywalking
👉🏻 https://github.com/apache/skywalking

#### Apache ShardingSphere
👉🏻  https://github.com/apache/shardingsphere

### Relevant articles

**CN:**
[Cyborg Flow X SkyWalking: Full Link Online Load Testing solution](https://skywalking.apache.org/zh/2022-01-18-cyborg-flow/)
[Full Link Online Load Testing for Production Database: Apache ShardingSphere Shadow Database Feature Upgrade](https://u01f1kqxrl.feishu.cn/docs/doccn3jWAIjs4jvjoSV7GiluqEc?appStyle=UI4&domain=www.feishu.cn&locale=zh-CN&refresh=1&tabName=space&theme=light&userId=6973852370656673795)

**EN:**
[Full Link Online Load Testing for Production Database: Apache ShardingSphere Shadow Database Feature Upgrade](https://blog.devgenius.io/full-link-online-stress-testing-for-production-database-apache-shardingsphere-shadow-database-84f7cba56f99)

**Comment by moonming** [comment]
Great，+1

**Comment by wu-sheng** [comment]
Here is what I propose

```
- cyborg-flow-gateway (APISIX extension, config file mostly)
- cyborg-agent (SkyWalking extension, new plugins for database supported by ShardingSphere)
- database-shield (ShardingSphere extension, config file and default settings)
```

Each folder should have its own structure and `Makefile` in its root path, and package all necessary binary(including upstream release and dev version) into `dist` folder.
Such as `cyborg-agent` should tar agent, oap backend, UI with new setup UI template in one place.

**Reply by tristaZero** [reply]
How about `database-shadow`？

**Reply by tristaZero** [reply]
Update my idea. I prefer `cyborg-database-shadow`

**Reply by wu-sheng** [reply]
Choose what you want. I just don't like feature included, because who knows what we will add in following years.

**Comment by wu-sheng** [comment]
@tristaZero Could you share what is the required comments in the SQL?

**Reply by tristaZero** [reply]
Sure, @soulasuna Please provide the comment format.

**Reply by soulasuna** [reply]
@tristaZero @wu-sheng @moonming 
I am very excited to participate in cooperation with everyone.
Basic example  `/*shadow:true,version:1.0,businessType:order*/ `
-  shadow:true ,   The default logo is like `CAFE BABE`.
-  version:1.0 ,    The test version identification.
-  businessType: order ,   The test business type identification.
Do you have any better suggestions?

**Reply by wu-sheng** [reply]
We need one flag for cyborg traffic.
How about we use cyborg-flow:true gloablly?

**Reply by soulasuna** [reply]
I agreed. This is a basic logo.

**Reply by soulasuna** [reply]
Like this  `/*cyborg-flow:true,version:1.0,business-type:order*/`.

**Reply by wu-sheng** [reply]
What do other two mean? I think one flag should be enough.

**Reply by soulasuna** [reply]
I think users may need it.
In the case of simultaneous testing of multiple versions or multiple business types.
`cyborg-flow:true` is basic flag.  The other two are extension.

**Comment by wu-sheng** [comment]
@mrproliu We need to implement a new MySQL plugin for CyborgFlow project. 

When context includes `cyborg-flow:true`, the plugin should inject `/*cyborg-flow:true */` at the end of raw SQL statements(prepared statement)

Please add the plugin to `{project.base}/cyborg-agent/plugins/mysql`.

**Reply by mrproliu** [reply]
Sure, what's that mean about the context? Or where is the context value from?

**Reply by wu-sheng** [reply]
I mean from here, https://skywalking.apache.org/docs/main/latest/en/protocols/skywalking-cross-process-correlation-headers-protocol-v1/.

This context should be injected by APISIX.
```
header-key: `sw8-correlation`
header-value: base64("cyborg-flow"):base64("true")
```
 @moonming This could be easily done through APISIX, right?

**Reply by moonming** [reply]
Yes, Apache APISIX can add HTTP headers by exist plugin.

**Reply by wu-sheng** [reply]
@moonming It would be great if your team to submit the configuration file or scripts in the `/cyborg-flow-gateway` to make APISIX adding this HTTP header in default.

**Reply by moonming** [reply]
> @moonming It would be great if your team to submit the configuration file or scripts in the `/cyborg-flow-gateway` to make APISIX adding this HTTP header in default.

@spacewander, please take a look

**Comment by tristaZero** [comment]
@moonming hasn’t feedbacked so far. How about start the cooperation between Skywalking and ShardingSphere? We’re almost ready for the next step.

**Reply by moonming** [reply]
I am back ...

**Reply by tristaZero** [reply]
Welcome. ;-)

**Reply by moonming** [reply]
@spacewander please take a look at this issue, IMO Apache APISIX can support this.

**Reply by spacewander** [reply]
Yes, APISIX can add the header for it.

**Comment by tristaZero** [comment]
@soulasuna Hi, could we provide a template shadow-configuration and guide to help @mrproliu setup?

**Reply by soulasuna** [reply]
@tristaZero OK.
@mrproliu Hi, reduce unnecessary configuration by users, shadow provides a unique streamlined configuration.
The configuration file and instructions will be delivered to the warehouse after the test is completed next Tuesday.

**Reply by mrproliu** [reply]
OK, I haven't started development yet. I can start writing only content related to skywalking, looking forward to your configuration. thanks:)

**Reply by soulasuna** [reply]
@mrproliu Hi, the `cyborg-database-shadow` project is created. For instructions, please refer to `README.md`. the template shadow-configuration in `cyborg-database-shadow/conf` directory.
Since ShardingSphere has not released the official version of 5.0.0, the project provided in `cyborg-database-shadow/server_src` can be used for development and testing. Just unzip it to use.
You can contact me if you have any questions.

**Reply by mrproliu** [reply]
Thanks, I will take a look tomorrow.

**Reply by mrproliu** [reply]
@soulasuna When I refer the `README.md` to build the `shardingsphere` project, I found one dependency could not be resolved, If I miss anything to setting?
missing dependency: `org.apache.shardingsphere:shardingsphere-integration-agent-test-common:jar:tests:5.0.0-RC1-SNAPSHOT`

**Reply by mrproliu** [reply]
I have used the shadow program to make proxy SQL requests, and it works! I will submit a PR tomorrow to provide the plugins, and I need to write some documents to explain before submitting.

**Reply by soulasuna** [reply]
@mrproliu Thanks, you can directly use the provided proxy project for development and testing in `server_src`. No need make, unzip and use.
